vector_search/client.cc

```diff
@@ -80,16 +102,45 @@ seastar::future<client::request_result> client::request(
         resp.content = co_await util::read_entire_stream(body);
     };
 
-    auto f = co_await seastar::coroutine::as_future(http_client_.make_request(std::move(req), std::move(handler), std::nullopt, &as));
-    if (f.failed()) {
-        auto err = f.get_exception();
-        co_return std::unexpected(co_await map_err(err));
-    }
+    co_await http_client_.make_request(std::move(req), std::move(handler), std::move(expected_status), &as);
     co_return resp;
 }
 
+seastar::future<bool> client::check_status() {
+    auto f = co_await coroutine::as_future(request_impl(httpd::operation_type::GET, "/api/v1/status", std::nullopt, http::reply::status_type::ok, as_));
+    auto ret = !f.failed();
+    co_await std::move(f).discard_result().handle_exception([](const auto&) {});
+    co_return ret;
+}
+
 seastar::future<> client::close() {
-    return http_client_.close();
+    as_.request_abort();
+    co_await gate_.close();
+    co_await http_client_.close();
+}
+
+seastar::future<> client::handle_server_unavailable() {
+    if (!is_checking_status_in_progress()) {
+        (void)try_with_gate(gate_, [this]() {
+            return run_checking_status().handle_exception([](const auto&) {});
+        });
+    }
+    co_return;
+}
+
+seastar::future<> client::run_checking_status() {
+    struct stop_retry {};
+    co_await exponential_backoff_retry::do_until_value(BACKOFF_RETRY_MIN_TIME, BACKOFF_RETRY_MAX_TIME, as_, [this] -> future<std::optional<stop_retry>> {
```

Comment:
I think that the discussion under the ticket suggested using 2 * operation timeout (the read timeout - `read_request_timeout_in_ms` - I suppose) as the upper bound. How hard would it be to carry this parameter here and use it? The parameter is live updateable, so ideally we should refer to it through `utils::updateable_value`.